### ISSUE TITLE
Forcing relay only on my clients and a port allocator results in never allocating a port

### ISSUE THREAD

**Issue by doctorpangloss** [pion/turn#229]
### Your environment.
  - Version: 72ffd594f7c6104b5680a4fea5651928e3b6ddb4
  - Browser: Mobile Safari on iOS 12.7, Chrome 92

### What did you do?
I added logging to `RelayAddressGeneratorPortRange` to see if a port is ever allocated.

```go
	for try := 0; try < r.MaxRetries; try++ {
		port := r.MinPort + uint16(r.Rand.Intn(int((r.MaxPort+1)-r.MinPort)))
		conn, err := r.Net.ListenPacket(network, fmt.Sprintf("%s:%d", r.Address, port))
		if err != nil {
		  log.Printf("could not allocate port %d", port)
			continue
		}

		relayAddr := conn.LocalAddr().(*net.UDPAddr)
		relayAddr.IP = r.RelayAddress
		log.Printf("allocated %d")
		return conn, relayAddr, nil
	}
```

### What did you expect?

Ports should be allocated.

### What happened?

A port is never allocated randomly within the range. I see multiple of the following errors:

```
turn ERROR: 2021/08/19 20:17:08 error when handling datagram: failed to handle Allocate-request from 192.168.88.162:61561: relay already allocated for 5-TUPLE
```

**Comment by adrianosela** [comment]
@doctorpangloss thank you for reporting.

I'm digging into this.

**Comment by adrianosela** [comment]
So, I'm not sure if this is the bug the reporter experienced or even a bug at all, but I was looking into the error message.

In `internal/server/turn.go`:

```
if alloc := req.AllocationManager.GetAllocation(fiveTuple); alloc != nil {
    id, attrs := alloc.GetResponseCache()
    if id != stunMsg.TransactionID {
        // reject with error
    }
}
```

**We reject allocations for 5-tuples already allocated if the transaction ID does not match the original request, even if the user is the same. I believe some clients may retry with different tx id(?) so perhaps we should consider returning successfully if the user is the same (despite the tx id mismatch). Unless this behavior is explicitly part of an RFC?**

To illustrate the point:

Say a client makes an allocation request:
- user: abc
- 5-tuple: (192.168.1.100:50000 -> 10.0.0.1:3478, UDP)
- tx id: 1111

And the allocation succeeds server side but the client thinks it failed (due to packet loss, etc), the client may retry with the same user and 5-tuple, but different tx id:
- user: abc
- 5-tuple: (192.168.1.100:50000 -> 10.0.0.1:3478, UDP)
- tx id: 2222

Currently our implementation replies to those retries with `437 Allocation Mismatch` errors. We should consider just checking if the allocation is for the same user and if so, returning the allocation details regardless of tx id.

**Comment by adrianosela** [comment]
Ah, nevermind. Per [RFC 5766 Section 6.2](https://datatracker.ietf.org/doc/html/rfc5766#section-6.2):

>   2.  The server checks if the 5-tuple is currently in use by an
>       existing allocation.  If yes, the server rejects the request with
>       a 437 (Allocation Mismatch) error.

and

> NOTE: When the Allocate request is sent over UDP, section 7.3.1 of
>      [RFC5389] requires that the server handle the possible
>      retransmissions of the request so that retransmissions do not
>     cause multiple allocations to be created.  Implementations may
>      achieve this using the so-called "stateless stack approach" as
>     follows.  To detect retransmissions when the original request was
>      successful in creating an allocation, the server can store the
>     transaction id that created the request with the allocation data
>    and compare it with incoming Allocate requests on the same
>   5-tuple.  Once such a request is detected, the server can stop
>      parsing the request and immediately generate a success response.
>     When building this response, the value of the LIFETIME attribute
>    can be taken from the time-to-expiry field in the allocate state
>      data, even though this value may differ slightly from the LIFETIME
>      value originally returned.  In addition, the server may need to
>      store an indication of any reservation token returned in the
>      original response, so that this may be returned in any
>      retransmitted responses.


TL;DR; current implementation is correct, per RFC.

**Comment by adrianosela** [comment]
Should this potentially be closed due to staleness @Sean-Der ?

**Comment by adrianosela** [comment]
What's the policy on stale issues cc:// @JoTurk - shall this be closed?
